### PR TITLE
Jupyter Lab Component Refactor

### PR DESCRIPTION
**Pull request by krishnakalyan3**
Support for Kernels
- [x] Python
- [x] R
- [x] Julia

### CODE REVIEW

File: lit_jupyter/component.py [hunk 1]
```diff
@@ -1,15 +1,63 @@
...
 import subprocess
+import sys
+import os
 import lightning as L
+from typing import Optional
+from pathlib import Path
 
-logger = logging.getLogger(__name__)
+
+pkg_install = """
```
Comment (zippeurfou) [inline lit_jupyter/component.py:10]: could this be in a file we read? Might be easier.

File: tests/test_lit_jupyter_component.py [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
 
-def test_placeholder_component():
-    component = LitJupyter()
-    # TODO: test that it's running
+@mock.patch("lit_jupyter.component.subprocess.Popen")
+def test_subprocess(mock_popen):
+    mock_popen().wait.return_value = 0
+    lightning_work = LitJupyter()
+    lightning_work.run()
+    return mock_popen.assert_called()
```
Comment (zippeurfou) [inline tests/test_lit_jupyter_component.py:10]: @tchaton wdyt of this test?

File: lit_jupyter/component.py [hunk 3]
```diff
@@ -1,15 +1,63 @@
...
+        # Start jupyter without password
+        remove_pwd = f"--NotebookApp.token='' --NotebookApp.password=''"
+        proc_server = subprocess.Popen(
+            f"{sys.executable} -m jupyter lab --ip {self.host} --port {self.port} --no-browser {remove_pwd}".split(" "),
+            stdout=jupyter_out,
+            stderr=jupyter_out,
+            bufsize=0,
+            close_fds=True
+        )
+        proc_server.wait()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply lit_jupyter/component.py:63]: Normally, we don't need to.

File: tests/test_lit_jupyter_component.py (hunk 2 above)
Comment (tchaton) [reply tests/test_lit_jupyter_component.py:10]: Yes, this is fine. But I would also check the arguments are correct.
This won't check that this component can be used properly locally and in the cloud. 
This is a mock test.

File: lit_jupyter/component.py [hunk 4]
```diff
@@ -1,15 +1,63 @@
...
 import lightning as L
+from typing import Optional
+from pathlib import Path
 
-logger = logging.getLogger(__name__)
+
+pkg_install = """
+sudo apt-get update
+sudo apt install r-base
+sudo R -e "install.packages('IRkernel')"
```
[the remark this answers is not in the document]
Comment (tchaton) [reply lit_jupyter/component.py:13]: I think you can do sudo within the container.

File: lit_jupyter/component.py [hunk 5]
```diff
@@ -1,15 +1,63 @@
...
+tar -xvzf julia-1.7.3-linux-x86_64.tar.gz; julia-1.7.3/bin/julia -e 'using Pkg; Pkg.add("IJulia")'
+rm -rf julia-1.7.3-linux-x86_64.tar.gz
+""".split('\n')
+
+
+class CustomBuildConfig(L.BuildConfig):
+    def build_commands(self):
+        return pkg_install
 
 class LitJupyter(L.LightningWork):
```
Comment (tchaton) [inline lit_jupyter/component.py:25]: I would call it JupyterLab directly.

File: lit_jupyter/component.py [hunk 6]
```diff
@@ -1,15 +1,63 @@
...
-logger = logging.getLogger(__name__)
+
+pkg_install = """
+sudo apt-get update
+sudo apt install r-base
+sudo R -e "install.packages('IRkernel')"
+Rscript -e "IRkernel::installspec()"
+wget https://julialang-s3.julialang.org/bin/linux/x64/1.7/julia-1.7.3-linux-x86_64.tar.gz
+tar -xvzf julia-1.7.3-linux-x86_64.tar.gz; julia-1.7.3/bin/julia -e 'using Pkg; Pkg.add("IJulia")'
+rm -rf julia-1.7.3-linux-x86_64.tar.gz
```
Comment (tchaton) [inline lit_jupyter/component.py:17]: Might be worth providing JupyterLab(kernel="julia|python|r|...")

File: lit_jupyter/component.py [hunk 7]
```diff
@@ -1,15 +1,74 @@
...
+        build_config = CustomBuildConfig(kernel)
+        super().__init__(cloud_compute=cloud_compute, cloud_build_config=build_config, parallel=True)
 
     def run(self):
-        remove_pwd = "--NotebookApp.token='' --NotebookApp.password=''"
-        host_info = f"--ip={self.host} --port={self.port}"
-        cmd = f"jupyter-lab --allow-root --no-browser {host_info} {remove_pwd}"
-        subprocess.run(cmd, shell=True)
\ No newline at end of file
+        jupyter_out = open('jupyter_log.txt', 'a')
```
Comment (tchaton) [inline lit_jupyter/component.py:42]: Don't forget to close the file.

File: demo_app.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+from lit_jupyter import JupyterLab
+
+class JupyterLabManager(L.LightningFlow):
+    def __init__(self) -> None:
+        super().__init__()
+        self.jupyter_work = JupyterLab(cloud_compute=L.CloudCompute(os.getenv("COMPUTE", "cpu-small")), kernel="python|r|julia")
+
+    def run(self):
+        self.jupyter_work.run()
+        if os.environ.get('TEST'):
```
Comment (tchaton) [inline demo_app.py:14]: You aren't testing that the component is actually working here. Check again my ModelServer Tutorial.

**Comment by krishnakalyan3** [comment]
cc: @manskx can I have a review from you?

File: demo_app.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python3
+
+import os
+import lightning as L
+from lit_jupyter import JupyterLab
+
+class JupyterLabManager(L.LightningFlow):
+    def __init__(self) -> None:
+        super().__init__()
+        self.jupyter_work = JupyterLab(kernel=os.getenv("KERNEL", "python"), cloud_compute=L.CloudCompute(os.getenv("COMPUTE", "cpu-small")))
```
Comment (manskx) [inline demo_app.py:10]: The `KERNEL` and `COMPUTE` are very generic and could cause some conflicts with other vars. I would recommend renaming the env vars something like `LIGHTNING_JUPYTER_LAB_KERNEL` `LIGHTNING_JUPYTER_LAB_COMPUTE`

File: README.md [hunk 10]
```diff
@@ -1,35 +1,66 @@
...
+# Installation
+Use these instructions to install:
+
+```
+lightning install component lightning/LAI-Jupyter-Component
+```
+
+Or to build locally
+```bash
+git clone https://github.com/PyTorchLightning/LAI-Jupyter-Component
```
Comment (manskx) [inline README.md:54]:
```suggestion
git clone https://github.com/Lightning-AI/LAI-Jupyter-Component

```

File: README.md [hunk 11]
```diff
@@ -1,35 +1,66 @@
...
-This ⚡ [Lightning component](lightning.ai) ⚡ was generated automatically with:
+A Lightning component to Launch Jupyter Lab
+______________________________________________________________________
 
-```bash
-lightning init component lit_jupyter
-```
-
-## To run lit_jupyter
+![Tests](https://github.com/PyTorchLightning/LAI-Jupyter-Component/actions/workflows/ci-testing.yml/badge.svg)
```
Comment (manskx) [inline README.md:7]:
```suggestion
![Tests](https://github.com/Lightning-AI/LAI-Jupyter-Component/actions/workflows/ci-testing.yml/badge.svg)
```

File: tests/test_jupyter_app.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+            print("Test successful, Exiting")
+            self._exit()
+
+    def configure_layout(self):
+        return {"name": "TestServer", "content": self.jupyter_work}
+
+
+def test_file_server():
+    app = L.LightningApp(RootFlow())
+    SingleProcessRuntime(app).dispatch()
```
Comment (manskx) [inline tests/test_jupyter_app.py:41]: Can't it use the `MultiProcessRuntime` ? The single process is kinda deprecated.

**Review by manskx — APPROVED** [review]

**Pull request merged**